lib/KBox/src/pages/NavigationPage.cpp

```diff
@@ -0,0 +1,280 @@
+                    apModeString = "APMode: Dodge     ";//extra spaces needed to overwrite old string if longer
+                    apModeDisplay->setText(apModeString);
+                    apModeDisplay->setColor(ColorRed);
+                }
+                return true;
+            } else {
+                DEBUG("apMode not active so no effect");
+            }
+        } else if (this->buttonPressedTimer > 5000) {  // Long click - start and stop the navMode here
+            DEBUG("long click");
+            if (imuCalibrated == true){
+                if (apMode == false){
+                    apMode = true;
+                    apHeadingMode = true;
+                    apModeString = "APMode: Heading       ";//extra spaces needed to overwrite old string if longer
+                    apModeDisplay->setText(apModeString);
+                    apModeDisplay->setColor(ColorGreen);
+                } else {
+                    apMode = false;
+                    apDodgeMode = false;
+                    apModeString = "APMode: Off           ";
+                    apModeDisplay->setText(apModeString);
+                    apModeDisplay->setColor(ColorBlue);
+                }
+            }
+            return true;
+        }
+    }
+    //DEBUG("APmode is: - %d", apMode);
+    //DEBUG("APHeadingmode is: - %d", apHeadingMode);
+    //DEBUG("APDodgemode is: - %d", apDodgeMode);
+    return true;
+}
+
+bool NavigationPage::processEvent(const EncoderEvent &ee) {
+    encoderClicks = ee.rotation;
+    navEncoderClicks = navEncoderClicks + encoderClicks;
+    encoderClicks = 0;
+    return true;
+}
+
+String NavigationPage::formatMeasurement(float measure, const char *unit) {
+  // extra spaces at the end needed to clear previous value completely
+  // (we use a non-fixed width font)
+  char s[10];
+  snprintf(s, sizeof(s), "%.0f %s  ", measure, unit); //do not display decimals for headings
+  return String(s);
+}
+
+
+void NavigationPage::processMessage(const KMessage &message) {
+  message.accept(*this);
+}
+
+void NavigationPage::visit(const APMessage &ap) {
+    navTargetRudderPosition = ap.getTargetRudderPosition();
+    navRudderCommandSent = ap.getRudderCommandSent();
+}
+
+void NavigationPage::visit(const RUDMessage &rm) {
+    navRudderSensorPosition = rm.getRudderDisplayAngle();
+}
+
+void NavigationPage::visit(const IMUMessage &imu) {
+    if (imu.getCalibration() == 3){
+        imuCalibrated = true;
+    }
+    if (imuCalibrated){
+        if (!apMode){
+            apModeString = "APMode: Off   ";
+            apModeDisplay->setText(apModeString);
+            apModeDisplay->setColor(ColorGreen);
+        }
+    } else {
+        apModeString = "Calibrating   ";
+        apModeDisplay->setText(apModeString);
+        apModeDisplay->setColor(ColorRed);
+    }
+    
+    rawHeading = RadToDeg(imu.getCourse());
+    if (rawHeading < 0){
+        navCurrentHeading = 720 + rawHeading;//adding 720 for nav math purposes
```

Comment:
This part of the code is pretty hard to understand and will be hard to debug.

Some ideas to make things easier here:
 - Write a little utility function `formatAngle()` that takes an angle in radian and returns a value in degrees between 0 and 360. Then reuse this everywhere you want to display an angle. You could have two if you want one that returns something between 0 and 360 and another one that returns values between -180 and +180.
 - Don't do navigation math in the navigation page. If the navigation page is not displayed then this code will not get executed. The autopilot needs to do all the navigation math. (If you follow my advice above to remove `navCurrentHeading` from `NAVMessage` then you will not need that math here and this function will just have to worry about displaying things.
 - Use only radians for the navigation math. You should only convert radians to degrees when displaying to the user.